### ISSUE TITLE
Native continuous batching always reads host logits, even when the request could be sampled entirely on device

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1022]
## Why the native seam returns host `Vec<Vec<f32>>`, and what the D2H actually buys

@justinchuby asked on #1015 whether returning host rows rather than a tensor forces a device copy. It does — 593.5 KB/step, measured and reported by the gate. But the return type is a symptom, not the cause, and the interesting part is what forces it.

### The cause is where the sampler runs, not the type

`ContinuousBatchManager::advance_row` does:

```rust
let mut logits = row.pending_logits.take()?;
let token_id = select_next_token_with_rng(&mut logits, &context, &row.options, &row.chain, &mut row.state.rng);
if let (Some(top_logprobs), Some(logprobs)) = (row.options.top_logprobs, row.state.logprobs.as_mut()) {
    logprobs.push(logprob_for_token(&logits, token_id, top_logprobs));
}
```

Note `&mut logits`: the `ProcessorChain` **rewrites the row in place**, and `logprob_for_token` reads it again afterwards. The seam returns a mutable host f32 buffer because that is what the sampler consumes. Returning a device tensor would not remove the copy; it would move it to wherever the chain runs.

The device-argmax fast path — which returns tokens and never moves logits — is preserved precisely for the case where none of that is needed.

### We already sample on device. The split is documented

`crates/onnx-genai-ort/src/device_sampler.rs` (`CudaSampler`) states it:

> The sampler applies the device-portable pipeline — temperature, top-k, top-p, min-p, then greedy (argmax) or categorical selection — **entirely on the device pointer, copying back only the 4-byte token id(s)**. History-dependent processors (repetition/frequency/presence penalties, grammar constraints, stop sequences) and logprobs **remain host-side**; `copy_row_to_host` serves those by copying the full row on demand.

So the 593.5 KB/step is not the price of sampling. It is the price of **history-dependent processors and logprobs**, which need the whole vocabulary on the host.

### The concrete gap

That device sampler is wired into the ORT captured-decode path (`onnx-genai-ort/src/decode`). The native `ContinuousBatchManager` path added in #1015 always takes host rows, whether or not the request needs them.

For a request whose options use only the device-portable processors and no `top_logprobs`, we could sample on device and copy back **4 bytes instead of 593.5 KB** — about a 150,000x reduction in bytes moved per row per step. For a request using repetition penalty or logprobs, the full row is genuinely required and the current path is correct.

That is the same shape as two decisions already made in this repo: keep both paths and choose per capability, rather than replacing one with the other. It is how the device-argmax fast path survives alongside host logits (#1009), and how the governed f32 weight cache is admitted or declined by budget (#987).

### What I am not claiming

I have not measured what fraction of the 593.5 KB/step actually costs throughput. At 32.8 ms of D2H over 13 steps on qwen0.5b it is visible but not obviously dominant, and on a larger model the per-step compute grows while the vocabulary does not. The win is real in bytes; whether it is material in wall-clock needs measuring on a quiet machine, and every timing taken here today was under heavy contention.

Filing as a follow-up rather than folding it into #1015, which is already validated and merged: mid-flight admissions at steps 2 and 7 through the real manager, all rows byte-identical to solo, `captures=1 replays=11 fallbacks=0 invalidations=0`, and the manager adding no copy on top of the read.

**Comment by justinchuby** [comment]
Read-only reconnaissance on main (`11043a02`), so this is buildable from a spec rather than a re-investigation. Three findings, one of which changes the design.

## 1. The capability predicate already exists and is not shared

`crates/onnx-genai-engine/src/decode_loop.rs:169-195` already computes exactly the "can this request be sampled on device" question, twice:

```rust
let greedy_fastpath = chain.is_empty()
    && options.top_logprobs.is_none()
    && (options.greedy || options.temperature == 0.0)
    && state.custom_sampler.is_none()
    && backend.greedy_fastpath_supported();

let sampled_fastpath = !greedy_fastpath
    && !options.greedy
    && options.temperature != 0.0
    && options.top_logprobs.is_none()
    && state.custom_sampler.is_none()
    && is_device_portable_chain(chain)
    && backend.sampled_fastpath_supported();
```

It is written inline in the single-request loop and reachable from nowhere else. `batched.rs` never consults it. So the work is not "invent a predicate", it is "lift this one into a named, testable function and call it per row". Lifting it is also worth doing on its own: a predicate this load-bearing that exists only as a local `let` in one function will drift the moment a second caller appears -- and this issue is that second caller.

## 2. The mixed batch is the whole problem, and the obvious implementation loses most of the win

This is the part that differs from the single-request case and the part I would most expect an implementation to get wrong.

One model step produces logits for **all** rows (`step_active` / `step_select` in `batched.rs:502-541`), then demuxes per row:

```rust
row.pending_logits = Some(take_row_logits(&mut logits, active_index, 0)?);   // line 511
row.pending_logits = Some(take_row_logits(&mut logits, row.physical_row, 0)?); // line 545
```

Rows in one batch have **independent** `options` and `ProcessorChain`s -- that is the point of continuous batching. So a batch is routinely mixed: some rows device-portable, some using repetition penalty or `top_logprobs`.

The tempting implementation is "if any active row needs host logits, take the host path for the batch." That is correct but nearly worthless: with independent requests arriving continuously, the probability that *every* concurrent row is device-portable falls off fast with batch size. One row asking for `top_logprobs` would tax every other row in the batch.

The design that actually pays is **per-row routing off one device-resident logits buffer**: device-sample the qualifying rows (4 bytes back each) and `copy_row_to_host` only the rows that genuinely need the vocabulary. `DeviceSampler::copy_row_to_host` (`device_sampler.rs:91`) already does exactly that single-row copy -- it was built for this shape.

So the seam change is: the batched step must be able to hand back a *device pointer plus row stride*, not an already-copied host `Value`. Today the D2H is inside `step_active` / `step_select` and the manager only ever sees host rows, which is why the return type looked like the cause.

## 3. What to assert, given a GPU is not available to me

Do not gate this on wall-clock. Every timing on this box today was contended, and I have no CUDA runtime installed, so I cannot referee a throughput claim.

Assert **bytes moved**, with a counter, the way `DIRECT_OUTPUT_CALLS` / `STAGED_OUTPUT_CALLS` prove routing in `qlinear_matmul.rs`. `logits_d2h_stats()` already exists at `batched.rs:339` and is the natural place to hang it. The test that matters is a **mixed batch**: N rows where some qualify and some do not, asserting that the copied bytes equal `(non-qualifying rows) x vocab x 4` and not `(all rows) x vocab x 4`. That test fails today, fails again under the naive whole-batch fallback, and passes only under per-row routing -- which is precisely the property worth locking in.

And per the standing rule: falsify it. Revert the routing and confirm the test fails.

## What I am still not claiming

Same caveat as the original post: the win is real in bytes (593.5 KB -> 4 B per qualifying row per step) and unmeasured in wall clock. A bytes-moved test is honest about that; a speedup claim would not be.
